core/src/main/scala/org/apache/spark/shuffle/PushShuffleWriterComponent.scala

```diff
@@ -0,0 +1,464 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.shuffle
+
+import java.io.File
+import java.net.ConnectException
+import java.nio.ByteBuffer
+import java.util.concurrent.ExecutorService
+
+import scala.collection.mutable.{ArrayBuffer, HashMap, HashSet, Queue}
+
+import com.google.common.base.Throwables
+
+import org.apache.spark.{ShuffleDependency, SparkConf, SparkEnv}
+import org.apache.spark.annotation.Since
+import org.apache.spark.internal.Logging
+import org.apache.spark.internal.config._
+import org.apache.spark.launcher.SparkLauncher
+import org.apache.spark.network.buffer.{FileSegmentManagedBuffer, ManagedBuffer, NioManagedBuffer}
+import org.apache.spark.network.netty.SparkTransportConf
+import org.apache.spark.network.shuffle.BlockFetchingListener
+import org.apache.spark.network.shuffle.ErrorHandler.BlockPushErrorHandler
+import org.apache.spark.network.util.TransportConf
+import org.apache.spark.shuffle.PushShuffleWriterComponent._
+import org.apache.spark.storage.{BlockId, BlockManagerId, ShufflePushBlockId}
+import org.apache.spark.util.{ThreadUtils, Utils}
+
+/**
+ * Used for pushing shuffle blocks to remote shuffle services when push shuffle is enabled.
+ * When push shuffle is enabled, it is created after the shuffle writer finishes writing the shuffle
+ * file and initiates the block push process.
+ *
+ * @param dataFile         mapper generated shuffle data file
+ * @param partitionLengths array of shuffle block size so we can tell shuffle block
+ *                         boundaries within the shuffle file
+ * @param dep              shuffle dependency to get shuffle ID and the location of remote shuffle
+ *                         services to push local shuffle blocks
+ * @param partitionId      map index of the shuffle map task
+ * @param conf             spark configuration
+ */
+private[spark] class PushShuffleWriterComponent(
```

Comment:
I'm ok to make it a separate class instead of an interface(thus, we can avoid initializing many fields when push shuffle is disabled).  But `XXXComponent` just sounds like a per executor instance to me, which could be used for multiple tasks. Although, it's actually per task. So, I personally would like to rename it to `ShuffleBlockPusher`, which is more consistent with `ShuffleBlockFetcherIterator`.

Reply:
I have renamed it to `ShuffleBlockPusher`.